qiskit/circuit/library/arithmetic/piecewise_polynomial_pauli_rotations.py

```diff
@@ -0,0 +1,260 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=no-member
+
+"""Piecewise-polynomially-controlled rotation."""
+
+from typing import List, Optional
+import numpy as np
+
+from qiskit.circuit import QuantumRegister
+from qiskit.circuit.exceptions import CircuitError
+
+from qiskit.circuit.library.arithmetic.functional_pauli_rotations import FunctionalPauliRotations
+from qiskit.circuit.library.arithmetic.polynomial_pauli_rotations import PolynomialPauliRotations
+from qiskit.circuit.library.arithmetic.integer_comparator import IntegerComparator
+
+
+class PiecewisePolynomialPauliRotations(FunctionalPauliRotations):
+    r"""Piecewise-polynomially-controlled Pauli rotations.
+
+    For a piecewise polynomial (not necessarily continuous) function :math:`f(x)`, which is defined
+    through breakpoints and coefficients as follows.
+    Suppose the breakpoints :math:`(x_0, ..., x_J)` are a subset of :math:`[0, 2^n-1]`, where
+    :math:`n` is the number of state qubits. Further on, denote the corresponding coefficients by
+    :math:`[a_{j,1},...,a_{j,d}]`, where :math:`d` is the highest degree among all polynomials.
+    Then :math:`f(x)` is defined as:
+
+    .. math::
+
+        f(x) = \begin{cases}
+            0, x < x_0 \\
+            \sum_{i=0}^{i=d}a_{j,i} x^i, x_j \leq x < x_{j+1}
+            \end{cases}
+
+    where we implicitly assume :math:`x_{J+1} = 2^n`.
+    """
+
+    def __init__(self,
+                 num_state_qubits: Optional[int] = None,
+                 breakpoints: Optional[List[int]] = None,
+                 coeffs: Optional[List[List[float]]] = None,
```

Comment:
Would it make sense to have this in a dictionary instead of a list? 
```python
coeffs = [
    [1, 2, 0],
    [-1],
    [0, 0, 0, 2]
]
PiecewisePolynomialPauliRotations(4, breakpoints=[0, 2, 3], coeffs=coeffs)
```
or 
```python
coeffs = {0: [1, 2, 0],
          2: [-1],
          3: [0, 0, 0, 2]}
PiecewisePolynomialPauliRotations(4, coeffs=coeffs)
```
Does numpy/scipy have a similar functionality? How do they do it?